# 배추값 예측
- 농산물 유통 정보 : https://www.kamis.or.kr/customer/main/main.do
- 날씨 정보 : https://data.kma.go.kr/stcs/grnd/grndRnList.do?pgmNo=69

## url의 이미
<pre>
    action=daily&
    startday=2023-06-01&   -> 시작일자
    endday=2023-06-04&     -> 종료일자
    countycode=&
    itemcategorycode=200   -> 배추
    &itemcode=211&         -> 배추
    kindcode=&
    productrankcode=&
    convert_kg_yn=N  
</pre>

In [3]:
import requests
from bs4 import BeautifulSoup
import datetime
import pandas as pd

In [4]:
# url = 'https://www.kamis.or.kr/customer/price/retail/period.do?action=daily&startday=2021-01-01&endday=2021-01-01&countycode=&itemcategorycode=200&itemcode=211&kindcode=&productrankcode=&convert_kg_yn=N'
url = "https://www.kamis.or.kr/customer/price/wholesale/period.do?action=daily&startday=2022-01-01&endday=2022-01-5&countycode=&itemcategorycode=200&itemcode=211&kindcode=&productrankcode=&convert_kg_yn=N"
response = requests.get(url)
html = BeautifulSoup(response.text, 'html.parser')
# price = int(html.select_one('table.wtable3 td.last').text.replace(',',''))
price = int(html.select_one('table.wtable3 td.last').text.replace(',',''))
print(price)
9270

9270


9270

In [5]:
# try:  시도 > 되면 식1 실행
#     식1   
# except:  안되면 식2 실행
#     식2   
try:
    response = requests.get(url)

    html = BeautifulSoup(response.text, 'html.parser')
    price = int(html.select_one('table.wtable3 td.last').text.replace(',',''))
    print(price)
except:
    pass

9270


In [6]:
str(datetime.date(2023,5,1))

'2023-05-01'

In [7]:
datetime.timedelta(days=1)

datetime.timedelta(days=1)

In [8]:
datetime.date(2023,5,1) + datetime.timedelta(days=1)

datetime.date(2023, 5, 2)

In [9]:
sd = datetime.date(2022,1,1)
ed = sd
final_ed = datetime.date(2022,12,31)

total_price = []
while final_ed != ed:
    try:
        url = f'https://www.kamis.or.kr/customer/price/retail/period.do?action=daily&startday={str(sd)}&endday={str(sd)}&countycode=&itemcategorycode=200&itemcode=211&kindcode=&productrankcode=&convert_kg_yn=N'
        response = requests.get(url)

        html = BeautifulSoup(response.text, 'html.parser')
        price = int(html.select_one('table.wtable3 td.last').text.replace(',',''))
        total_price.append([sd,price])
    except:
        pass
    
    sd += datetime.timedelta(days=1)
    ed = sd
    print(ed, end=" ")
df = pd.DataFrame(total_price, columns = ['날짜', '배추가격'])

2022-01-02 2022-01-03 2022-01-04 2022-01-05 2022-01-06 2022-01-07 2022-01-08 2022-01-09 2022-01-10 2022-01-11 2022-01-12 2022-01-13 2022-01-14 2022-01-15 2022-01-16 2022-01-17 2022-01-18 2022-01-19 2022-01-20 2022-01-21 2022-01-22 2022-01-23 2022-01-24 2022-01-25 2022-01-26 2022-01-27 2022-01-28 2022-01-29 2022-01-30 2022-01-31 2022-02-01 2022-02-02 2022-02-03 2022-02-04 2022-02-05 2022-02-06 2022-02-07 2022-02-08 2022-02-09 2022-02-10 2022-02-11 2022-02-12 2022-02-13 2022-02-14 2022-02-15 2022-02-16 2022-02-17 2022-02-18 2022-02-19 2022-02-20 2022-02-21 2022-02-22 2022-02-23 2022-02-24 2022-02-25 2022-02-26 2022-02-27 2022-02-28 2022-03-01 2022-03-02 2022-03-03 2022-03-04 2022-03-05 2022-03-06 2022-03-07 2022-03-08 2022-03-09 2022-03-10 2022-03-11 2022-03-12 2022-03-13 2022-03-14 2022-03-15 2022-03-16 2022-03-17 2022-03-18 2022-03-19 2022-03-20 2022-03-21 2022-03-22 2022-03-23 2022-03-24 2022-03-25 2022-03-26 2022-03-27 2022-03-28 2022-03-29 2022-03-30 2022-03-31 2022-04-01 2022-04-02

In [10]:
df.to_csv('ml_datas/cabbage/배추가격_202201_202212.csv', encoding='utf-8', index=False)

In [40]:
df = pd.read_csv('ml_datas/cabbage/배추가격_202201_202212.csv')
df.head(10)

날짜  배추가격
0  2022-01-03  4151
1  2022-01-04  4289
2  2022-01-05  4302
3  2022-01-06  4387
4  2022-01-07  4387
5  2022-01-10  4326
6  2022-01-11  4489
7  2022-01-12  4506
8  2022-01-13  4559
9  2022-01-14  4330

In [41]:
df_temp = pd.read_csv('ml_datas/cabbage/2022_기온.csv', skiprows=7, encoding='euc-kr')
df_temp.drop('지점', axis=1, inplace=True)
df_temp['날짜'] = df_temp['날짜'].str.replace('\t', '')
df_temp.head()

날짜  평균기온(℃)  최저기온(℃)  최고기온(℃)
0  2022-01-01     -3.7     -9.9      3.4
1  2022-01-02     -0.8     -6.7      6.1
2  2022-01-03     -1.5     -7.8      4.8
3  2022-01-04      0.0     -5.0      5.2
4  2022-01-05     -0.8     -6.0      4.0

In [42]:
df_rain = pd.read_csv('ml_datas/cabbage/2022_강수량.csv', skiprows=7, encoding='euc-kr')
df_rain.drop('지점', axis=1, inplace=True)
df_rain.head()

날짜  강수량(mm)
0  2022-01-01      0.0
1  2022-01-02      0.1
2  2022-01-03      0.0
3  2022-01-04      0.0
4  2022-01-05      0.2

In [43]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   날짜       365 non-null    object 
 1   평균기온(℃)  365 non-null    float64
 2   최저기온(℃)  365 non-null    float64
 3   최고기온(℃)  365 non-null    float64
dtypes: float64(3), object(1)
memory usage: 11.5+ KB


In [44]:
df_rain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   날짜       365 non-null    object 
 1   강수량(mm)  365 non-null    float64
dtypes: float64(1), object(1)
memory usage: 5.8+ KB


In [45]:
df_temp_rain = pd.merge(df_temp, df_rain, on='날짜')
df_temp_rain.head()

날짜  평균기온(℃)  최저기온(℃)  최고기온(℃)  강수량(mm)
0  2022-01-01     -3.7     -9.9      3.4      0.0
1  2022-01-02     -0.8     -6.7      6.1      0.1
2  2022-01-03     -1.5     -7.8      4.8      0.0
3  2022-01-04      0.0     -5.0      5.2      0.0
4  2022-01-05     -0.8     -6.0      4.0      0.2

In [46]:
df.shape

(246, 2)

In [47]:
df_temp_rain.shape

(365, 5)

In [48]:
data = pd.merge(df_temp_rain, df, on='날짜', how='inner')
data

날짜  평균기온(℃)  최저기온(℃)  최고기온(℃)  강수량(mm)  배추가격
0    2022-01-03     -1.5     -7.8      4.8      0.0  4151
1    2022-01-04      0.0     -5.0      5.2      0.0  4289
2    2022-01-05     -0.8     -6.0      4.0      0.2  4302
3    2022-01-06      0.0     -5.0      6.2      0.0  4387
4    2022-01-07     -1.6     -7.7      6.7      0.0  4387
..          ...      ...      ...      ...      ...   ...
241  2022-12-26     -3.5     -8.8      3.1      0.0  2955
242  2022-12-27     -2.3     -8.2      4.8      0.0  2955
243  2022-12-28     -1.6     -5.7      3.3      0.0  2972
244  2022-12-29     -2.5     -7.5      3.0      0.0  3000
245  2022-12-30     -1.2     -5.7      4.0      0.0  3011

[246 rows x 6 columns]

In [49]:
data.set_index('날짜', inplace=True)
data.head()

평균기온(℃)  최저기온(℃)  최고기온(℃)  강수량(mm)  배추가격
날짜                                                  
2022-01-03     -1.5     -7.8      4.8      0.0  4151
2022-01-04      0.0     -5.0      5.2      0.0  4289
2022-01-05     -0.8     -6.0      4.0      0.2  4302
2022-01-06      0.0     -5.0      6.2      0.0  4387
2022-01-07     -1.6     -7.7      6.7      0.0  4387

## 머신러닝 모델 생성

In [51]:
## EDA 필요시 분할 전 먼저 해야 한다

### 데이터 분할

In [50]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np

In [52]:
# X
X = data.drop(["배추가격"], axis=1)
X

평균기온(℃)  최저기온(℃)  최고기온(℃)  강수량(mm)
날짜                                            
2022-01-03     -1.5     -7.8      4.8      0.0
2022-01-04      0.0     -5.0      5.2      0.0
2022-01-05     -0.8     -6.0      4.0      0.2
2022-01-06      0.0     -5.0      6.2      0.0
2022-01-07     -1.6     -7.7      6.7      0.0
...             ...      ...      ...      ...
2022-12-26     -3.5     -8.8      3.1      0.0
2022-12-27     -2.3     -8.2      4.8      0.0
2022-12-28     -1.6     -5.7      3.3      0.0
2022-12-29     -2.5     -7.5      3.0      0.0
2022-12-30     -1.2     -5.7      4.0      0.0

[246 rows x 4 columns]

In [55]:
y = data["배추가격"]
y

날짜
2022-01-03    4151
2022-01-04    4289
2022-01-05    4302
2022-01-06    4387
2022-01-07    4387
              ... 
2022-12-26    2955
2022-12-27    2955
2022-12-28    2972
2022-12-29    3000
2022-12-30    3011
Name: 배추가격, Length: 246, dtype: int64

In [58]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2021)

In [59]:
from xgboost import XGBRegressor

In [60]:
# n_estimators : 학습기 사용 갯수
# learning_rate : 건너 뛰는 갯수

xgb_reg = XGBRegressor(n_estimators = 500, learning_rate = 0.05)

In [61]:
xgb_reg.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.05, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=500, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

### 모델 테스트

In [62]:
pred = xgb_reg.predict(X_test)

In [63]:
mse = mean_squared_error(y_test, pred)
rmse = np.sqrt(mse)
print(rmse)

1216.4592910326273


In [ ]:
# 2023-06-07
# 평균기온 : 21.2 최저기온 : 16.3  최고기온 : 26.4 강수량 : 0
# 배추(10kg): 6,548 , 

In [64]:
temp_avg = float(input('평균기온 :'))
temp_min = float(input('최저기온 :'))
temp_max = float(input('최고기온 :'))
rain = float(input('강수량 :'))

평균기온 : 21.2
최저기온 : 16.3
최고기온 : 26.4
강수량 : 0


In [73]:
test_df = pd.DataFrame({'평균기온(℃)':[temp_avg], '최저기온(℃)': [temp_min],
                   '최고기온(℃)': [temp_max] ,'강수량(mm)': [rain] })

test_df.values

array([[21.2, 16.3, 26.4,  0. ]])

In [74]:
rain = test_df.values
rain

array([[21.2, 16.3, 26.4,  0. ]])

In [75]:
X.columns

Index(['평균기온(℃)', '최저기온(℃)', '최고기온(℃)', '강수량(mm)'], dtype='object')

In [76]:
test_df = pd.DataFrame(rain, columns=X.columns, dtype='float')

In [77]:
test_df

평균기온(℃)  최저기온(℃)  최고기온(℃)  강수량(mm)
0     21.2     16.3     26.4      0.0

### 배추 가격 예측

In [78]:
xgb_reg.predict(test_df)

array([4411.594], dtype=float32)

#### 2023-06-07
- 실제 가격
    - 8246

- ML 이 예측한 배추 가격
    - 평균 가격 : 4411.594
    - 오차 : 1216.46

## r2 score(결정 계수) 확인

- r2 score 값이 1에 수렴할수록 좋은 모델 (X는 y를 잘 설명하는 데이터)
- r2 score 값이 0에 수렴할수록 나쁜 모델 (X는 y를 잘 설명하는 데이터가 아님)
- r2 score 값이 -1 사용 불가

In [80]:
from sklearn.metrics import r2_score

In [83]:
xgb_test_pred = xgb_reg.predict(X_test)

In [84]:
print("r2_score: {}".format(r2_score(y_test, xgb_test_pred)))

r2_score: 0.20631645018234857
